### ISSUE TITLE
The mIoU of split0 is lower than that in the paper

### ISSUE THREAD

**Issue by fangzy97** [Reagan1311/ASGNet#3]
Thank for sharing your code. When I run the training by
`sh tools/train.sh pascal asgnet split0_resnet50` 
with your default setting and run the testing by
`sh tools/test.sh pascal asgnet split0_resnet50`
using the checkpoint `final.pth`, I get the miou of 0.5389 which is much lower than 58.84 in the paper. Do I miss some steps ?

**Comment by Reagan1311** [comment]
Hi, thanks for raising the question. The model will provide two pth files after training. 
One is the `final.pth`, and the other one is `train_epoch_xxx.pth`, which means the checkpoint with the best result. 
Generally, the final checkpoint doesn't represent the best performance, so you should use the checkpoint of the other one.

**Comment by fangzy97** [comment]
Thank for your reply. I get the result using `train_epoch_70.pth`. I'm still a little confused about the code. In the `train.py` the `val_data` is set by:

`val_data = dataset.SemData(split=args.split, shot=args.shot, max_sp=args.max_sp, data_root=args.data_root,
                                              data_list=args.val_list, transform=val_transform, mode='val',
                                              use_coco=args.use_coco, use_split_coco=args.use_split_coco)`

and in the `test.py` the `val_data` is set by:

`val_data = dataset.SemData(split=args.split, shot=args.shot, max_sp=args.max_sp, data_root=args.data_root,
                                              data_list=args.val_list, transform=val_transform, mode='val',
                                              use_coco=args.use_coco, use_split_coco=args.use_split_coco)`

It seems that they use the same parameters. Is that means you are using the `test set` as the `validation set` to choose the best model or I miss something ?

**Comment by Reagan1311** [comment]
Yeah, that's a general setting in few-shot segmentation.

**Comment by fangzy97** [comment]
Thanks, it's really helpful for me.

**Comment by meetviolet** [comment]
Hi, could anyone help me bcz my experiment server is low and I set **data loader workers** to be 0 or 1 or 4, and I found it always stuck at 26th print out.
![image](https://user-images.githubusercontent.com/25797747/227600002-f24f11af-94c1-4958-81d7-a646f9f12d53.png)
![image](https://user-images.githubusercontent.com/25797747/227600071-7b7dd2ec-3236-430f-8396-2307815d8841.png)


And I want to know how can get the train_epoch_70.pth or could anyone share it to me?
/(ㄒoㄒ)/~~My graduation subject is so urgency and I really need pth model to output something.

**Comment by fangzy97** [comment]
May be you can try to do some experiments on the pascal voc dataset instead of the coco.

**Comment by Reagan1311** [comment]
If you GPU is low on memory, you can reduce the input size or batch size, though this may sacrifice the accuracy.

**Comment by meetviolet** [comment]
Hi, I ran Pascal dataset and it seems runing successfully but I can't use multi gpu to run code.

I set 
train_gpu: [0,1] and batch_size=2 in
/ASGNet/data/config/pascal/pascal_asgnet_split0_resnet50.yaml

and set
os.environ["CUDA_VISIBLE_DEVICES"] = '0,1'
in train.py

But it still run on only one gpu.

Do u know how to make multi gpu to run?

**Comment by fangzy97** [comment]
Actually, the `train_gpu` in the config file does not make sense and I just set `train_gpu: [0]`. In my experiments, I modified `train.sh` to the following content.
```
PARTITION=Segmentation

dataset=$1
model_name=$2
exp_name=$3
gpus=$4

if [ "x${gpus}" = "x" ];then
  gpus=0
fi

exp_dir=exp/${model_name}/${dataset}/${exp_name}
model_dir=${exp_dir}/model
result_dir=${exp_dir}/result
config=data/config/${dataset}/${dataset}_${model_name}_${exp_name}.yaml

mkdir -p ${model_dir} ${result_dir}
now=$(date +"%Y%m%d_%H%M%S")
cp tool/train.sh tool/train.py model/${model_name}.py ${config} ${exp_dir}

CUDA_VISIBLE_DEVICES=${gpus} python -u -m tool.train --config=${config} 2>&1 | tee ${result_dir}/train-$now.log

```

Then, the command `./tool/train.sh pascal asgnet split0_resnet50 [gpu_ids]` to start a experiment. For example, if you want to use the GPU:0 and GPU:1, the command should be
```
./tool/train.sh pascal asgnet split0_resnet50 0,1
```
Note that if you use my script, you should comment out `os.environ["CUDA_VISIBLE_DEVICES"] = '0,1'` in `train.py`.

**Comment by meetviolet** [comment]
Thank you very much♥♥♥(ಥ _ ಥ), and I would like to ask that
The multi-GPU running experiment, is the image dataset allocated to multi Gpus?
Or just use the whole dataset on every gpu and balance the final performance parameters？

**Comment by fangzy97** [comment]
Science the implementation use the `nn.DataParallel` to parallel the model, you do not need to modify the code in the multi-GPU running experiment.
